### DISCUSSION TITLE
Perfomance comparison Oxigraph vs. QLever

### DISCUSSION THREAD

**Discussion by hannahbast** [oxigraph/oxigraph#841] [General]
Dear Thomas,

I finally found the time to play around with Oxigraph a bit. Let me first say that I am super impressed with the extent and the professionalism of this project and this repository. All the more since so far this has essentially been a one-person project. It's a great idea to provide the various components (like the RDF parser or the SPARQL parser) as independent modules/crates. Compilation from source was unproblematic and the command-line interface is easy to use and self-explanatory. In particular, it is very easy to load a dataset and start a server. Everything just works. In the world of academe, this is the absolute exception.

I compared loading time, index size, and query time of Oxigraph vs. QLever for a moderately sized RDF dataset, namely https://dblp.org/rdf/dblp.ttl.gz (1.7 GB compressed, 390 M triples), and a variety of queries (see below). Everything was run on an AMD Ryzen 9 7950X 16-Core machine with 128 GB and 7.1 TB of NVMe SSD (high-quality but affordable consumer hardware, total cost around 2500 €).

Loading time was 640s for Oxigraph (0.6 M triples/sec) vs. 231s for QLever (1.7 M triples/sec) on NVMe SSD. On HDD, it was 2537s for Oxigraph vs. 270s for QLever (apparently, Oxigraph makes heavy use of random access during loading). The total size of the index files was 66.5 GB for Oxigraph vs. 7.7 GB for QLever (apparently, Oxigraph doesn't compress much yet). I am curious whether the proportions of these stats carry over to a larger dataset like Wikidata (19 B triples). For QLever, load time and index size are essentially proportional to the size of the input dataset.

Here are the results for six queries from https://qlever.cs.uni-freiburg.de/dblp ("Examples"), selected for their variety. For QLever, the cache was cleared before each query. For Oxigraph, no special precautions were taken, except that the server was started from scratch once at the beginning. Both servers were run on SSD. For Oxigraph, it can make a huge difference when the disk cache is empty (`sudo bash -c "sync; sleep 5; echo 3 > /proc/sys/vm/drop_caches"`). For the queries strongly affected by this, I indicated this by writing X -> Y, where X is the query time with empty disk cache and Y is the query time when repeating the query.

| Query | Oxigraph | QLever | Result shape | Comment / Question |
| -------|---------|------------|-------------|------------|
| [All papers published in SIGIR](https://qlever.cs.uni-freiburg.de/dblp/m3osZX) | 1.5s -> 0.3s | 0.01s | 6254 x 3 | Two simple joins, nothing special |
| [Number of papers by venue](https://qlever.cs.uni-freiburg.de/dblp/weowtF) | 2.6s -> 1.8s | 0.02s | 19954 x 2 | Scan of a single predicate with GROUP BY and ORDER BY |
| [Author names matching REGEX](https://qlever.cs.uni-freiburg.de/dblp/bVZBoH) | 5.5s -> 0.5s | 0.03s | 513 x 3 | Joins, GROUP BY, ORDER BY, FILTER REGEX |
| [All papers in DBLP until 1940](https://qlever.cs.uni-freiburg.de/dblp/Rd9ixQ) | 310s | 0.10s | 70 x 4 | Three joins, a FILTER, and an ORDER BY, why does Oxigraph take so long? |
| [All papers with their title](https://qlever.cs.uni-freiburg.de/dblp/WzABwD) | 134s -> 83s | 4.16s | 716786 x 2 | Single predicate, but must materialize large result (problematic for many SPARQL engines) |
| [All predicates ordered by the number of subjects](https://qlever.cs.uni-freiburg.de/dblp/8R9a3u) | 102s | 0.01s | 68 x 2 | Conceptually requires a scan over all triples, but huge optimization potential |

**Comment by hannahbast** [comment]
P.S. In the meantime, I have included other engines in the comparison (it's now: Oxigraph, Apache Jena, Stardog, Blazegraph, Virtuoso, QLever). Results are reported here: https://github.com/ad-freiburg/qlever/wiki/QLever-performance-evaluation-and-comparison-to-other-SPARQL-engines . The first table gives a nice overview.

**Comment by joepie91** [comment]
@hannahbast In light of this post, I decided to try running [my scraping server](https://codeberg.org/joepie91/srap-next) against QLever - it is a somewhat unusual usecase for RDF stores, it seems, as it uses the store much more like a 'traditional' database, in that it continuously updates its contents (starting from nothing) and just treats the store as a persistent database. So it's much less query-heavy and much more update/mutation-heavy.

I found QLever to perform very poorly in this usecase; I did not take quantitative measurements, but in my output logs I could see [tasks for seekseek](https://codeberg.org/seekseek/config-v2) being completed one by one whereas normally with Oxigraph some of those tasks scroll across the screen too fast to read - so a very superficial estimate would be that QLever is at least an order of magnitude slower than Oxigraph in this usecase.

I realize that this isn't how people usually use RDF stores (or at least, it seems that they don't), so it may well be out of scope for QLever, but it may perhaps be worth investigating as another usage pattern to benchmark?

**Comment by Jimflip** [comment]
I use QLever for a project and is very good and improving, for large mostly static datasets, highly recommend over things like Stardog.

I'm using Oxigraph for my current project, the size of the graphs are much smaller but importantly it's more operational, triples regularly being added and many graphs etc.  It has been surprisingly robust and sufficiently performant and convenient.

Both stores are good but using the right tool for the project is key!

**Comment by hannahbast** [comment]
@joepie91 I missed your comment, and just stumbled upon this very old thread, prompted by @Jimflip's comment. Since 0.5.45 (released on Feb 24) QLever can rebuild its index internally and since https://github.com/ad-freiburg/qlever/pull/3163, these rebuilds can be triggered automatically when the number of updates exceeds a certain threshold; see the description of that PR:

That way, you can now start QLever also with an empty dataset and update it as much as you want. The index will periodically rebuild itself automatically and answer queries fast. No server restarts needed, no queries are getting lost. Without these periodic rebuilds and when starting with an empty dataset, QLever operates without its index data structures and the performance is terrible.
